### PR TITLE
Kafka ACLs support

### PR DESCRIPTION
**Pull request by BenjaminDebotte**
# Extract of application.conf
```
    # spark jobs
    "kafka-topic-1" {
      accessControlEntries = [
        {
          principal ="User"
          name = "bdebotte"
          hosts = ["*"]
          operations = [
            "READ",
            "WRITE"
          ]
          permissions = [
            "ALLOW"
          ]
        }
      ]
    },
    "kafka-topic-1" {
      accessControlEntries = [
        {
          principal ="User"
          name = "admin"
          hosts = ["*"]
          operations = [
            "READ",
            "WRITE",
            "CREATE"
          ]
          permissions = [
            "DENY"
          ]
        }
      ]
    },
```

### CODE REVIEW

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala [hunk 1]
```diff
@@ -32,6 +33,11 @@ object TopicSettings {
     TopicSettings(
       config.as[Int]("rf"),
       config.as[Int]("partitions"),
+      if (config.hasPath("accessControlEntries")) {
+        config.as[Array[TopicAclEntry]]("accessControlEntries")
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala]: `config.getAs[Seq[TopicAclEntry]].getOrElse(Seq.empty)`

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettingsOverrides.scala [hunk 2]
```diff
@@ -23,24 +23,27 @@ import scala.collection.JavaConverters._
...
 
 object TopicSettingsOverrides {
-  val empty = TopicSettingsOverrides(None, None, Map.empty)
+  val empty = TopicSettingsOverrides(None, None, None, Map.empty)
 
   implicit val valueReader: ValueReader[TopicSettingsOverrides] = ValueReader.relative { config =>
     TopicSettingsOverrides(
       config.getAs[Int]("rf"),
       config.getAs[Int]("partitions"),
+      config.getAs[Array[TopicAclEntry]]("accessControlEntries"),
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettingsOverrides.scala]: kebab-case

**Review by ilya-epifanov — CHANGES_REQUESTED** [review]

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala [hunk 3]
```diff
@@ -15,6 +15,10 @@
 
 package com.libertyglobal.odh.kafkatool
 
+
+import java.util
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: excusez-moi?

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala [hunk 4]
```diff
@@ -267,12 +272,31 @@ object Main extends StrictLogging {
...
+        applyAcls(kafka, config)
+
       case _ =>
         opts.printHelp()
         sys.exit(1)
     }
   }
 
+  private def listAcls(kafka: AdminClient): Unit = {
+    ACLManager.list(kafka).foreach(acl => println(acl.toString))
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: println -> logger.info

File: src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        kafka.createAcls(acls.asJavaCollection).all().get()
+      }
+      true
+    } catch {
+      case ee: ExecutionException => false
+    }
+  }
+
+  def list(kafka: AdminClient): Seq[AclBinding] = {
+        list(kafka, AclBindingFilter.ANY)
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala]: please tick the "Reformat code" checkbox in commit dialog

File: src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    res.values().get().asScala.toSeq
+  }
+
+  def delete(kafka: AdminClient, aclFilters: Seq[AclBindingFilter] ): Seq[AclBinding] = {
+    val res = kafka.deleteAcls(aclFilters.asJavaCollection)
+    res.all().get().asScala.toSeq
+  }
+
+  def deleteAll(kafka: AdminClient): Seq[AclBinding] = {
+    delete(kafka, Seq() :+ AclBindingFilter.ANY)
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala]: Seq(AclBindingFilter.ANY) ?

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        case "WRITE" => AclOperation.WRITE
+        case "CREATE" => AclOperation.CREATE
+        case "DELETE" => AclOperation.DELETE
+        case "ALTER" => AclOperation.ALTER
+        case "DESCRIBE" => AclOperation.DESCRIBE
+        case "CLUSTER_ACTION" => AclOperation.CLUSTER_ACTION
+        case "DESCRIBE_CONFIGS" => AclOperation.DESCRIBE_CONFIGS
+        case "ALTER_CONFIGS" => AclOperation.ALTER_CONFIGS
+        case "IDEMPOTENT_WRITE" => AclOperation.IDEMPOTENT_WRITE
+        case _ => AclOperation.UNKNOWN
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: please just fail the operation

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      }
+    }
+
+    def strToPermission(p: String): AclPermissionType = {
+      p.toUpperCase match {
+        case "ALLOW" => AclPermissionType.ALLOW
+        case "DENY" => AclPermissionType.DENY
+        case "ANY" => AclPermissionType.ANY
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: no need to support ANY either, we're only writing ACLs here

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        case _ => AclPermissionType.UNKNOWN
+      }
+    }
+
+    val kafkaPrincipal = new KafkaPrincipal(principal, name)
+
+    for (host <- hosts;
+         operation <- operations;
+         permission <- permissions)
+      yield new AccessControlEntry(kafkaPrincipal.toString, host, strToOperation(operation), strToPermission(permission))
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: I'd write this as:

```scala
for {
  host <- hosts
  operation <- operations
  permission <- permissions
} yield new AccessControlEntry(kafkaPrincipal.toString, host, strToOperation(operation), strToPermission(permission))
```

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+
+}
+
+object TopicAclEntry {
+  implicit val valueReader: ValueReader[TopicAclEntry] = ValueReader.relative { config =>
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: this shouldn't be needed as all the fields have supported types and simple names

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala [hunk 11]
```diff
@@ -32,6 +33,7 @@ object TopicSettings {
     TopicSettings(
       config.as[Int]("rf"),
       config.as[Int]("partitions"),
+      if (config.hasPath("access-control-entries")) config.as[Seq[TopicAclEntry]]("access-control-entries") else Seq.empty,
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala]: config.getAs

File: src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+object ACLManager {
+
+  def add(kafka: AdminClient, acls: Seq[AclBinding], dryRun: Boolean = false): Boolean = {
+    try {
+      if (!dryRun) {
+        kafka.createAcls(acls.asJavaCollection).all().get()
+      }
+      true
+    } catch {
+      case ee: ExecutionException => false
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala]: please never ever swallow exceptions

File: src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.kafka.clients.admin.AdminClient
+import org.apache.kafka.common.acl.{AclBinding, AclBindingFilter}
+
+import scala.collection.JavaConverters._
+
+object ACLManager {
+
+  def add(kafka: AdminClient, acls: Seq[AclBinding], dryRun: Boolean = false): Boolean = {
+    try {
+      if (!dryRun) {
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala]: I'd apply `dryRun` logic at the caller site

File: src/main/scala/com/libertyglobal/odh/kafkatool/Opts.scala [hunk 14]
```diff
@@ -41,5 +41,14 @@ class Opts(arguments: Seq[String]) extends ScallopConf(arguments) {
   val listSuperfluousTopics = new Subcommand("list-superfluous-topics")
   addSubcommand(listSuperfluousTopics)
 
+
+  val acl_run = new Subcommand("acl-run") {
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Opts.scala]: maybe `update-acls`?

Comment (ilya-epifanov) [reply src/main/scala/com/libertyglobal/odh/kafkatool/Opts.scala]: and please use `snakeCase` for member names

**Review by ilya-epifanov — CHANGES_REQUESTED** [review]

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala (hunk 11 above)
Comment (BenjaminDebotte) [reply src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala]: I tried what you suggested yesterday and that, nothing shows up nor seems to work

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala [hunk 15]
```diff
@@ -34,24 +36,89 @@ import scala.reflect.io.File
 
 object Main extends StrictLogging {
 
-  def sortedToString[T: Ordering](xs: Iterable[T]): String = {
-    xs.toSeq.sorted.mkString(",")
-  }
+  def main(args: Array[String]): Unit = {
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: how come that the whole file is changed now?

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala (hunk 11 above)
Comment (ilya-epifanov) [reply src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicSettings.scala]: TopicSettingsOverrides.scala:46

**Review by ilya-epifanov — CHANGES_REQUESTED** [review]

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala (hunk 15 above)
Comment (BenjaminDebotte) [reply src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: I made the `Reformat code` thing, haven't noticed it's a mess. I'll arrange it

File: .gitignore [hunk 16]
```diff
@@ -1,3 +1,5 @@
 target/
 sbt/*.jar
 sbt/*.jar.part
+
+\.idea/
```
Comment (ilya-epifanov) [inline .gitignore]: this really should be in your global .gitignore

File: src/main/scala/com/libertyglobal/odh/kafkatool/Opts.scala [hunk 17]
```diff
@@ -41,5 +41,14 @@ class Opts(arguments: Seq[String]) extends ScallopConf(arguments) {
   val listSuperfluousTopics = new Subcommand("list-superfluous-topics")
   addSubcommand(listSuperfluousTopics)
 
+
+  val aclUpdate = new Subcommand("update-acls") {
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Opts.scala]: why not `val updateAcls` ?

File: src/main/scala/com/libertyglobal/odh/kafkatool/Opts.scala [hunk 18]
```diff
@@ -41,5 +41,14 @@ class Opts(arguments: Seq[String]) extends ScallopConf(arguments) {
...
+    val dryRun = opt[Boolean]("dry-run", 'n', descr = "Don't change anything")
+  }
+  addSubcommand(aclUpdate)
+
+  val aclList = new Subcommand("list-acls")
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Opts.scala]: here as well

File: src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala [hunk 19]
```diff
@@ -0,0 +1,32 @@
+package com.libertyglobal.odh.kafkatool.aclmanager
+
+import org.apache.kafka.clients.admin.AdminClient
+import org.apache.kafka.common.acl.{AclBinding, AclBindingFilter}
+
+import scala.collection.JavaConverters._
+
+object ACLManager {
+
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala]: unnecessary blank line

File: src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+  def deleteAll(kafka: AdminClient): Seq[AclBinding] = {
+    delete(kafka, Seq(AclBindingFilter.ANY))
+  }
+
+  def delete(kafka: AdminClient, aclFilters: Seq[AclBindingFilter]): Seq[AclBinding] = {
+    val res = kafka.deleteAcls(aclFilters.asJavaCollection)
+    res.all().get().asScala.toSeq
+  }
+
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/aclmanager/ACLManager.scala]: unnecessary blank line

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala [hunk 21]
```diff
@@ -18,12 +18,31 @@ package com.libertyglobal.odh.kafkatool.config
...
+import org.apache.kafka.common.resource.{Resource, ResourceType}
 
 case class KafkaToolConfig(
                             kafka: Map[String, AnyRef],
                             topicSettings: Map[String, TopicSettings],
                             brokerIds: Set[Int]
-                          )
+                          ) {
+  def getAcls() : Seq[AclBinding] = {
+
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala]: unnecessary blank line

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+import net.ceedubs.ficus.readers.ValueReader
+import org.apache.kafka.common.acl.{AccessControlEntry, AclOperation, AclPermissionType}
+import org.apache.kafka.common.security.auth.KafkaPrincipal
+
+case class TopicAclEntry(principal: String, name: String,
+                    hosts: Array[String], operations: Array[String],
+                    permissions: Array[String]) {
+
+  def toAccessControlEntries(): Array[AccessControlEntry] = {
+
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: unnecessary blank line

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala (hunk 8 above)
Comment (ilya-epifanov) [reply src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: please remove `"ANY"` case

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala (hunk 9 above)
Comment (ilya-epifanov) [reply src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: please rewrite this, using `for {}` for multiline `for`s is considered best practice

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+    val kafkaPrincipal = new KafkaPrincipal(principal, name)
+
+    for (host <- hosts;
+         operation <- operations;
+         permission <- permissions)
+      yield new AccessControlEntry(kafkaPrincipal.toString, host, strToOperation(operation), strToPermission(permission))
+  }
+
+
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: unnecessary blank lines

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+  implicit val valueReader: ValueReader[TopicAclEntry] = ValueReader.relative { config =>
+    TopicAclEntry(
+      config.as[String]("principal"),
+      config.as[String]("name"),
+      config.as[Array[String]]("hosts"),
+      config.as[Array[String]]("operations"),
+      config.as[Array[String]]("permissions")
+    )
+  }
+}
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala]: EOL at EOF

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+
+import net.ceedubs.ficus.Ficus._
+import net.ceedubs.ficus.readers.ValueReader
+import org.apache.kafka.common.acl.{AccessControlEntry, AclOperation, AclPermissionType}
+import org.apache.kafka.common.security.auth.KafkaPrincipal
+
+case class TopicAclEntry(principal: String, name: String,
+                    hosts: Array[String], operations: Array[String],
+                    permissions: Array[String]) {
+
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/TopicAclEntry.scala:11]: unnecessary blank line

**Review by ilya-epifanov — CHANGES_REQUESTED** [review]

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala [hunk 26]
```diff
@@ -267,12 +269,34 @@ object Main extends StrictLogging {
...
   }
 
+  private def listAcls(kafka: AdminClient): Unit = {
+    ACLManager.list(kafka).foreach(acl => logger.info(acl.toString))
+  }
+
+  private def applyAcls(kafka: AdminClient, config: KafkaToolConfig, dryRun: Boolean = false): Unit = {
+    val acls = config.getAcls()
+    if (dryRun) {
+      ACLManager.list(kafka).foreach(acl => logger.info("To-be deleted ACL " + acl.toString))
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: I'd say 
```scala
for (acl <- ACLManager.list(kafka)) {
  logger.info(s"To-be deleted ACL ${acl}")
}
```
looks more concise

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala [hunk 27]
```diff
@@ -267,12 +269,34 @@ object Main extends StrictLogging {
...
+      acls.foreach((acl) => logger.info("To-be added ACL " + acl.toString))
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: same here

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala [hunk 28]
```diff
@@ -267,12 +269,34 @@ object Main extends StrictLogging {
...
+    } else {
+      ACLManager.deleteAll(kafka).foreach(acl => logger.warn("Deleted ACL " + acl.toString))
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: please rewrite `foreach` as `for` loops, as in previous branch

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala [hunk 29]
```diff
@@ -18,15 +18,32 @@ package com.libertyglobal.odh.kafkatool.config
...
+import org.apache.kafka.common.resource.{Resource, ResourceType}
 
 case class KafkaToolConfig(
                             kafka: Map[String, AnyRef],
                             topicSettings: Map[String, TopicSettings],
                             brokerIds: Set[Int]
-                          )
+                          ) {
+  def getAcls() : Seq[AclBinding] = {
+    def buildResource(topic: String): Resource = new Resource(ResourceType.TOPIC,topic)
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala]: formatting

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala [hunk 30]
```diff
@@ -18,15 +18,32 @@ package com.libertyglobal.odh.kafkatool.config
...
+
+    def buildAclBindings(topic: String, aclEntries: Seq[TopicAclEntry]): Seq[AclBinding] = {
+      aclEntries
+        .flatMap(aclEntry => aclEntry.toAccessControlEntries())
+        .map(ACE => new AclBinding(buildResource(topic), ACE))
+
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala]: unnecessary blank line

File: src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala [hunk 31]
```diff
@@ -18,15 +18,32 @@ package com.libertyglobal.odh.kafkatool.config
...
+    }
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/config/KafkaToolConfig.scala]: need a blank line after function definition

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala [hunk 32]
```diff
@@ -267,12 +269,34 @@ object Main extends StrictLogging {
         updateCommand(kafka, config, opts.update.alterIfNeeded.getOrElse(false), opts.update.dryRun.getOrElse(false))
       case Seq(c) if c == opts.listSuperfluousTopics =>
         listSuperfluousTopicsCommand(kafka, config)
+      case Seq(c) if c == opts.listAcls =>
+        listAcls(kafka)
+      case Seq(c) if c == opts.updateAcls =>
+        applyAcls(kafka, config, opts.update.dryRun.getOrElse(false))
```
Comment (ilya-epifanov) [inline src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: could you please rename it to `updateAcls` for consistency?

File: src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala (hunk 26 above)
Comment (ilya-epifanov) [reply src/main/scala/com/libertyglobal/odh/kafkatool/Main.scala]: `.foreach` is slightly less readable than `for` loop. Sometimes it's easier to "end the sentence" with `.foreach` than "start the sentence" with `for` but usually it's the case when you have a lot of `map`s, `flatmap`s and other operations in a long chain.

**Review by ilya-epifanov — CHANGES_REQUESTED** [review]

**Review by ilya-epifanov — APPROVED** [review]

**Pull request merged**